### ISSUE TITLE
We need 64-bit CCKD support!

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#20]
This is just a single ToDo item on my long Hercules To-Do list.

Basically I need to either carefully modify existing Compressed CKD (CCKD) code to allow dasd image and shadow files to grow larger than 4GB (which is the current limit due to the use of 32-bit variables (e.g. file offsets) instead of 64-bit variables), or else I need to create a copy of the existing code and carefully modify it instead (thus creating a new CCKD64 device type, but leaving existing CCKD support intact).

_**OR...**_ &nbsp; I need to design/develop brand new (from scratch) support for CCKD64 myself (since either way is going to cause an obvious incompatibility issue)

I'm currently leaning towards writing brand new code myself.

___
**Note:** Issue #167 "[CCKD64 support needed for cckddump/cckdload utils](https://github.com/SDL-Hercules-390/hyperion/issues/167)" is a sub-project (sub-issue) of this issue.

**Comment by Peter-J-Jansen** [comment]
A brand new CCKD64 device would be my preference allowing for a smooth migration where needed.  The current limit of 8 shadow files was never a problem for me.

 

Cheers,

Peter J.

**Comment by mhoes** [comment]
Well I don't pretend to have any say in the matter, or for that matter even pretend to know what I'm talking about, but...

To me it seems to make the most sense to create a new 64-bit device type, and leave the current 32-bit version intact (no specific preference if that's brand new code, or a modified version of the existing code, although re-using existing code may have less risk of introducing new bugs, by mostly using code that is known to run correctly ?). It just seems to be the 'safest' way to make sure existing installations don't break when Hyperion is updated ? (Could be wrong here). Of course, if the end-goal is to drop support for 32-bit Hyperion builds completely and migrate all the code to 64-bit builds only, keeping the 32bit version around does not really make sense anymore.

Ok, I'll crawl back under my rock now. ;)

**Comment by Fish-Git** [comment]
I'm going to use the renamed copy of the existing CCKD code approach, thereby leaving the existing code intact. It's the safest technique.

**Comment by Fish-Git** [comment]
I am temporarily stopping work on this issue in order to focus attention on issue #99 instead ("Poor performance of CLCL, MVCIN and TRT instructions"). Once issue #99 is resolved I shall return to working on this issue again.
&nbsp;
___
**Side bar:** While working on this issue I discovered evidence of a _possible bug in the current CCKD implementation:_ running `dasdcopy` multiple times (copying an uncompressed image to a compressed one) -- using the same input and run options -- yields _different results each time!_ That is to say, the resulting output file is _different_ each time `dasdcopy` is run using the exact same input file and run options!

What's odd is that, so far, it is always approximately at the exact same location in the output file too (and the difference is only a few bytes). I only just discovered this phenomenon today too, so I have no further details to offer as I have not had time yet to look into it any further (nor at this point do I even know _how_ to go about looking into it).

**Comment by Fish-Git** [comment]
**Side bar update:** I did a few more quick tests and the results are quite interesting (and somewhat concerning): when I do two `dasdcopy -lfs -o ckd infile outfile` runs back to back with each run specifying a different output file (so they can be compared with each other afterwards), the results appear to differ _depending on the size of the output file!_ (wtf?)

Images whose resulting output file is over 4GB in size appear to exhibit the unexpected behavior whereas images whose resulting output file is _less than_ 4GB appear to behave as expected:

        M01RES: (10020 cyls) different
        S1USS1: (10000 cyls) different
        S1BLZ2: ( 5500 cyls) different
        S1IMU2: ( 3600 cyls) same
        S1RES1: ( 3339 cyls) same

This only occurs with dasd that actually have data too. When I tried the same thing using empty dasd created via dasdinit (e.g. `dasdinit -z  -r -lfs outfile.cckd 3390 5060` which specifies a 5060 cylinder cckd compressed output file which, when decompressed via `dasdcopy` yields an uncompressed ckd file slightly larger than 4GB), it doesn't fail. The results from the two back-to-back runs are identical to one another. The problem only seems to occur when the input cckd compressed files are non-empty (have presumably a significant amount of data on them). I'm not sure what any of this means but it's definitely weird and definitely concerning.

I don't really have the time to look into this any further but thought I should at least make a record of my findings. I might need to create a new github issue for this so it can be looked into further (more deeply) at some point in the future.

Back to issue 99...

**Comment by Fish-Git** [comment]
FYI to those interested: I recently finished(?) coding CCKD64 support and have begun testing it and it is looking really good! It successfully passes my quick preliminary tests.

There are still _many_ much more thorough tests I still need to run of course, but things are looking very good for a potential preliminary experimental release in a few weeks. Hopefully even before Christmas if I'm lucky! &nbsp; `:-D`

**Comment by Fish-Git** [comment]
_**COMMITTED!**_&nbsp;&nbsp;`:))`

Please consider this to be an _**experimental beta release**_ until such time as those interested have had a chance to give it a thorough workout to shake out any bugs that might still remain.

Happy Holidays, everyone!&nbsp;&nbsp;`:))`

(I will keep this issue open to allow people to report any bugs they find to this GitHub Issue. Only after sufficient time has passed (a few weeks? _months?_) without any new bug reports being reported will I then close this issue as having _finally_ been resolved!)

**Comment by moshix** [comment]
Well by just creating minidisks alone it won't exceed 4GB (I think it's smart enough to not actually write every block, but just to the high level formatting, but can't be sure). But by writing into them (appending a line to a file) it does and it has. 

thanks

M

**Comment by ivan-w** [comment]
Currently running a simple test...

copying a file with fairly random data multiple times unto a 3390-9 sized minidisk... 
When CMS mindisk is full, I look at the size of the CCKD volume and check the copies files are identical to the original ;)

Should a be a good stress test !

**Comment by moshix** [comment]
Yes, simple enough. Let me code that and then restar the test. Will revert when it's been running for a good 12-16 hours. Thank you Ivan. 

M

**Comment by ivan-w** [comment]
Failing at this point... 

Getting : 

HHC00304E 0:0307 CCKD file[0] ../zvm63-dasds/ZVM307.CCKD: get space error, size exceeds 4096M

.. Lemme check I didn't mess up something somewhere else...

Fish... Do I need to do something to convert a CCKD to use 64 bit offsets ?

--Ivan

**Comment by ivan-w** [comment]
For reference : 
Hercules : 
HHC01413I Hercules version 4.1.0.0-SDL-g025dfcd4 (4.1.0.0)
HHC01414I (C) Copyright 1999-2018 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Dec  7 2018 at 18:36:29
HHC01417I Built with: GCC 6.3.0 20170516

dasdinit : 'I recreated the volume just to make sure)
s$ dasdinit -z ZVM307.CCKD 3390-9 WRK002
HHC01413I Hercules utility dasdinit - DASD image file creation program; version 4.1.0.0-SDL-g025dfcd4 (4.1.0.0)
HHC01414I (C) Copyright 1999-2018 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Dec  7 2018 at 18:36:29
HHC00462I 0:0000 CKD file ZVM307.CCKD: creating 3390 volume WRK002: 10017 cyls, 15 trks/cyl, 56832 bytes/track
HHC00460I 0:0000 CKD file ZVM307.CCKD: 10017 cylinders successfully written
HHC02423I DASD operation completed

Still won't allow volumes above 4G and generate a HHC00304E error when the volume exceeds 4GB.. problems ensue (z/VM fails to shutdown trying to commit MINIDISK cache, etc...)

--Ivan

**Comment by Fish-Git** [comment]
> It'd be interesting to see if the volumes exceed 4GB and if the content written is the same as the content read !

I already tested that! That was part of my set of tests that I did before releasing it.

I wrote a standalone program(*) (using Harold's excellent ASMA/SATK(**)) that writes 6GB of random uncompressible data of a fixed pattern (the first 8 bytes of which is the count field) and then reads it back in again and verifies what was read is what was written and it works fine. (And yes, the actual resulting size _was_ over 6GB in size!)

Then I created a shadow file and used the same program to write a _different_ random pattern of uncompressible data, reading it back in and verifying its correctness. Then I did a backwards merge, merging the 6GB shadow (with pattern 2) into the 6GB compressed based (that had pattern 1), and read _that_ back in to verify it was all correct pattern 2.

I also did a series of sfk and sfc, dasdcopy64 with/without shadows, cckdcdsk64, cckddiag64, cckdcomp64, etc... Basically everything I could think of to throw at it, and it _**all**_ works!

I wouldn't release untested software of this importance, Ivan! Sheesh!&nbsp;`:)`
___
(*) Available to anyone interested.
(**) https://github.com/s390guy/SATK

**Comment by Fish-Git** [comment]
> Fish... Do I need to do something to convert a CCKD to use 64 bit offsets ?

_(Oops!)_ Sorry about that! I suppose I should create a README or something for CCKD64, eh? Yes, you obviously need to do some things different of course. You need to use either `dasdinit64` to _create_ a cckd64 dasd image, or else use `dasdcopy64` to _copy_ a cckd image into a cckd64 image. (Enter `dasdinit64` or `dasdcopy64`, etc, with no arguments for help.)

The first 8 bytes of the dasd image should be either CKD_C064 for a cckd64 compressed image or CKD_S064 for a compressed cckd64 shadow file. If they don't have that signature(*) then they're not cckd64 images and yes, you _will_ then run into the HHC00304E "size exceeds 4096M" problem.
___
(*) Refer to the new "[Compressed Dasd Emulation](https://sdl-hercules-390.github.io/html/cckddasd.html)" emulation web page for detailed information.

**Comment by Fish-Git** [comment]
Some of the utilities don't have a '64' variant and work fine with both cckd _-or-_ cck64 images. (They can detect which type of file they're working with and automatically call the correct support function appropriately.)

Some utilities however (such as `dasdinit64` and `dasdcopy64`, etc) required writing a separate '64' version (as they cannot so easily be made "cckd format indifferent"(*)) and you need to be careful to use the '64' version where/when appropriate.

I'll try to throw together a README...

___
(*) That is my eventual goal however. I fully recognize having two different versions of some utilities (dasdcop, dasdcopy64, dasdinit, dasdinit64, etc) is confusing. It'd be better if each utility handled either/both formats (cckd vs. cckd64), but as I said, for some utilities, the effort involved in making that change was a bit more than I was prepared to tackle at this point in time. I was more focused on getting it out the door than making it pretty! Merging dasdcopy64 into dasdcopy (and dasinit64 into dasdinit) can come later.

**Comment by moshix** [comment]
I am testing with a  completely re-installed z/VM 6.3 on top of 3390-3. All the DASDs are CCKD and initialized with the new hercules tools, so far still no issues. 

thanks

M

**Comment by moshix** [comment]
Fish,

One oddity I find with z/VM 6.3 is when I format a compressed 3390-3 (yes, initalized with dasdinit64) it is very slow at formatting for cylinders below 3900 and then gets faster. Look at this from the cpformatxa command:

FORMATTING OF CYLINDER 100 ENDED AT: 06:07:28  
FORMATTING OF CYLINDER 200 ENDED AT: 06:07:57  
FORMATTING OF CYLINDER 300 ENDED AT: 06:08:21  
FORMATTING OF CYLINDER 400 ENDED AT: 06:08:46  
FORMATTING OF CYLINDER 500 ENDED AT: 06:09:11  
FORMATTING OF CYLINDER 600 ENDED AT: 06:09:39  
FORMATTING OF CYLINDER 700 ENDED AT: 06:10:02  
FORMATTING OF CYLINDER 800 ENDED AT: 06:10:24  
FORMATTING OF CYLINDER 900 ENDED AT: 06:10:39  
FORMATTING OF CYLINDER 1000 ENDED AT: 06:10:48 
FORMATTING OF CYLINDER 1100 ENDED AT: 06:10:54 
FORMATTING OF CYLINDER 1200 ENDED AT: 06:10:59 
FORMATTING OF CYLINDER 1300 ENDED AT: 06:11:06 
FORMATTING OF CYLINDER 1400 ENDED AT: 06:11:12 
FORMATTING OF CYLINDER 1500 ENDED AT: 06:11:17 
FORMATTING OF CYLINDER 1600 ENDED AT: 06:11:22 
FORMATTING OF CYLINDER 1700 ENDED AT: 06:11:27 
FORMATTING OF CYLINDER 1800 ENDED AT: 06:11:31 
FORMATTING OF CYLINDER 1900 ENDED AT: 06:11:37 
FORMATTING OF CYLINDER 2000 ENDED AT: 06:11:42 
FORMATTING OF CYLINDER 2100 ENDED AT: 06:11:48 
FORMATTING OF CYLINDER 2200 ENDED AT: 06:11:52 
FORMATTING OF CYLINDER 2300 ENDED AT: 06:11:57 
FORMATTING OF CYLINDER 2400 ENDED AT: 06:12:02 
FORMATTING OF CYLINDER 2500 ENDED AT: 06:12:10 
FORMATTING OF CYLINDER 2600 ENDED AT: 06:12:14 
FORMATTING OF CYLINDER 2700 ENDED AT: 06:12:20 
FORMATTING OF CYLINDER 2800 ENDED AT: 06:12:25 
FORMATTING OF CYLINDER 2900 ENDED AT: 06:12:31 
FORMATTING OF CYLINDER 3000 ENDED AT: 06:12:35 
FORMATTING OF CYLINDER 3100 ENDED AT: 06:12:41 
FORMATTING OF CYLINDER 3200 ENDED AT: 06:12:46 
FORMATTING OF CYLINDER 3300 ENDED AT: 06:12:52 
FORMATTING OF CYLINDER 3400 ENDED AT: 06:12:57 
FORMATTING OF CYLINDER 3500 ENDED AT: 06:13:04 
FORMATTING OF CYLINDER 3600 ENDED AT: 06:13:09 
FORMATTING OF CYLINDER 3700 ENDED AT: 06:13:17 
FORMATTING OF CYLINDER 3800 ENDED AT: 06:13:22 
FORMATTING OF CYLINDER 3900 ENDED AT: 06:13:27 
FORMATTING OF CYLINDER 4000 ENDED AT: 06:13:33 
FORMATTING OF CYLINDER 4100 ENDED AT: 06:13:38 
FORMATTING OF CYLINDER 4200 ENDED AT: 06:13:45 
FORMATTING OF CYLINDER 4300 ENDED AT: 06:13:50 
FORMATTING OF CYLINDER 4400 ENDED AT: 06:13:55 
FORMATTING OF CYLINDER 4500 ENDED AT: 06:14:00 
FORMATTING OF CYLINDER 4600 ENDED AT: 06:14:06 

the time stamps show that it sometimes takes 22 seconds to format 100 cyinders but then gets much faster.  the system is otherwise idle. Real spinning disks (not SSD). 

thanks

M

**Comment by ivan-w** [comment]
With dasdinit64 - works as expected ;)

**Comment by Fish-Git** [comment]
> the time stamps show that it sometimes takes 22 seconds to format 100 cyinders but then gets much faster. the system is otherwise idle. Real spinning disks (not SSD). Never seen this behavior with 32bit CCKDs.

I have, and I _suspect_ the cckd logic's lock handling might be the culprit.

Existing cckd code (which of course was carried over to cckd64 code) involves heavy use of various internal locks, and I suspect the various internal helper threads that is spins off might be fighting with each other when things first start up, but then later get "out of sync" with each other such that they no longer are. This phenomenon (which is completely unconfirmed and is pure speculation at this point as I haven't had time to look into it yet) seem to be more likely to occur when there is lots of I/O being done to the same volume (same Hercules dasd image file).

By default, two "writer" threads are created to handle file writes and two "read-ahead" threads are created to perform anticipatory reading (which cache what they read so that the data is there waiting in the cache to synchronously satisfy the expected upcoming read I/O). Additionally, there is also a "garbage-collector" thread that periodically kicks in to try and compress updated dasd image files "on the fly" (to prevent them from excessive growth due to free space fragmentation when there are a lot of writes done to a file).

When a format is first started the threads, having all been started at almost the same time, are all "in sync" with one another, executing in "lock step" with one another, increasing the likelihood that they will be fighting for contention over the various internal locks used to coordinate access to internal structures. As time goes by however, their execution synchronicity "skews" from one another such that they're now "out of sync" with each other and no longer fighting over internal locks (or not fighting _as much_ (or as often)).

At least that's my theory right now.&nbsp;`:)`

If and when I have time I might try to look into this issue deeper to try and see what's actually going on, but it's understandably _very_ low priority right now.

If you have lots of spare processors (cores) on that monster box of yours (how many do you have?), you might want to, just for fun, play around with tweaking your `cckd` run options (e.g. the `wr`, `rat`, and `nostress` options (and/or any other options you may wish to experiment with).

Refer to the description of the cckd command/statement on Hercules's "[Compressed Dasd Emulation](https://sdl-hercules-390.github.io/html/cckddasd.html#cckdcommand)" web page for more information.

HTH!

**Comment by Fish-Git** [comment]
> With dasdinit64 - works as expected. `;-)`

_(Yawn!)_

&nbsp; &nbsp; &nbsp; `;-)`

**Comment by ivan-w** [comment]
Spurious message in dasdcopy : 
/   0%     185HHC90020W 'hthread_setschedparam()' failed at loc=cckddasd.c:1705: rc=22: Invalid argument

(Cosmetic only)

**Comment by Fish-Git** [comment]
> (I know.. I didn't RTFM !)

That was _my_ fault! I didn't update enough of the documentation to explain the need to use the new '64' versions of the various utilities. This was just corrected by a commit I made a few minutes ago.

**Comment by Fish-Git** [comment]
> `HHC90020W 'hthread_setschedparam()' failed at loc=cckddasd.c:1705: rc=22: Invalid argument`

Are running with elevated (root?) privileges? If not, then that's the cause. You need root privileges (Administrator privileges on Windows) in order to set priorities!

**Comment by ivan-w** [comment]
I think I was just expecting things to be done as Greg intended (Dynamic conversion.. smart but tricky).. 
No it was me - I should know better :P (you DID say it was beta - I should know there are quirks)
More reports to come, comments and whatnot !

**Comment by ivan-w** [comment]
About the HHC90020W message... You shouldn't need to be root to run a standalone utility to convert files you own to another set of files you also own. I think the message should only apply when cckdasd.c is running under hercules, not when running standalone. (and I think the message should be trashed together unless you have debug/trace on the dasd device)

**Comment by Fish-Git** [comment]
> I think the message should only apply when cckdasd.c is running under hercules, not when running standalone.

Agreed!

> (and I think the message should be trashed together unless you have debug/trace on the dasd device)

That's not a bad idea either!

**Comment by ivan-w** [comment]
When converting with dasdcopy/dasdconv, it seems the copy is decompressing and then compressing blocks again EVEN if blocks are using the same compression level/algorithm. Is this normal (or just a shortcut ?)

**Comment by Fish-Git** [comment]
> Is this normal (or just a shortcut ?)

Whatever it's doing is original design! I didn't change any of Greg's logic. I just copied it to another source file (e.g. `cckddasd64.c`) and made the changes to use the new cckd64 structs.

If you have a quality visual diff too (the one I used on Windows is ExamDiff Pro) you can compare each of the `*.64,c` source files with their non-64 counterparts too see the actual different between the two. The only differences there should be are to which structs are being used.

**Comment by ivan-w** [comment]
Running a complete z/VM 6.3 with converted dasds... I will monitor this and report !

**Comment by ivan-w** [comment]
But what happens when we reach the 16EB limit ? (Just kidding.. we should be safe)

**Comment by Peter-J-Jansen** [comment]
Fish,

I understand that both dasdcopy and dasdcopy64 support the optional sf=sfile parameter, but does this mean that when specified, an implicit merge is performed during the copy / conversion ?

Thanks,

Peter

**Comment by Fish-Git** [comment]
> I understand that both dasdcopy and dasdcopy64 support the optional sf=sfile parameter, but does this mean that when specified, an implicit merge is performed during the copy / conversion?

If that's what the old code did then that's what the new 64 code does too!

_The thing that each of you need to keep in mind is that I did **not** change any of the existing  cckd functionality!_

If the "old" code did or did not support something then the new cckd64 code does or does not support the same thing in the same way.

Thus if dasdcopy did an implicit merge when a shadow file was specified (optional sf=sfile parameter specified on the command line), then that's what the new cckd64 code will do too. I didn't change any of that!

(But to answer your question without even looking at the code, yes, I believe it _would_ do a merge!)

**Comment by Fish-Git** [comment]
> This time SDL latest completely freezes. I can't even switch from log view to hardware console view. Only kill -9 helps.

Well, that's not good!!&nbsp;`:(`

&nbsp;
> I could be wrong but looks like an SIE problem to me.

Sure looks that way to me too! &nbsp; The evidence for this, is the log message:

    HHC00822I Processor IE00: machine check due to host error: User defined signal 1

which tells me a SIGSEGV (or similar) occurred on CPU0 while in SIE mode (hence the " IE00" processor identification).

_**BUT**_... until there is evidence to the contrary, this doesn't appear to be a problem with cckd64 support, so I would suggest you create a new, separate GitHub Issue for this. That is to say, the problem should _**not**_ be pursued here. This thread/issue is for reporting problems related to cckd64 support. The fact that this problem occurred during cckd64 testing is immaterial IMO. Thus a new/separate GitHub Issue should be be opened/created for this problem so we can track it there.

Thanks.

**Comment by ivan-w** [comment]
May or may NOT be related... 

IFI004   
Explanation
HCPIFIIN was attempting to do one of the following:
Transfer control to the interrupt response address specified in the IORBK field IORIRA, but it contained zeros.
Call the emergency error recovery procedure, but was unable to resolve its address.
User response
For the first case, check R10 to see that it is pointing to a valid IORBK. If R10 contains the address of an IORBK, the interrupt response address was probably destroyed or never set when the request was initiated. Also examine the status in IORBK.
For the second case, verify that the emergency error recovery procedure (HCPIEERP) was included in the system.

Please examine using the dump using IPCS (Operatns)... It could be an invalid ERP in the sense too.

--Ivan

**Comment by moshix** [comment]
New issue found:

with SDL Latest when I create a 3390-9 DASD with dasdinit64 and then attach it to a running Hercules instance, z/VM sees it as 3390-3 (not CCKD 64). 

thanks

moshix

**Comment by Fish-Git** [comment]
> It might have something to do with the reuse of DEVBLKs.

Nope. &nbsp; That works fine too.

I have no idea what went wrong with you.

 Are you _sure_ you did what you _think_ you did?  Can we see some _evidence_ of this bug? Some logs of the commands you issued along with a Hercules log that shows the incorrect classification?

**Comment by moshix** [comment]
I am pretty sure. As you know I work on Linux. I am currently traveling to Europe but when I get there I will connect and just simply try again. I did try several times this morning and it went wrong each time. 

Give me about a day please 

Thanks 

M

**Comment by Fish-Git** [comment]
> As you know I work on Linux.

Your host being Linux shouldn't have anything to with it. CCKD code is platform independent.

> Give me about a day please

Take your time!

**Comment by Peter-J-Jansen** [comment]
Fish, thanks for your CCKD64 efforts !
I want to test the one case where CCKD64 support could become a lifesaver. Some of the z/OS 2.3 3390-9 DASD are verrry close to the 4GB limit, so much that I don't even dare running without shadow files. I successfully converted them all to CCKD64, like this : 

```
C:\Program Files\Hercules\R4.1>dasdcopy64 -z -o CCKD64 C:\Hercules\ZOS23\DASD\a3prd3.cckd sf=C:\Hercules\ZOS23\DASD\SHADOW\H2\a3prd3_2.cckd C:\Hercules\ZOS23\DASD64\a3prd3.cckd
HHC01413I Hercules utility dasdcopy64 - 64-bit DASD copy/convert; version 4.1.0.9449-SDL-g025dfcd4 (4.1.0.9449)
HHC01414I (C) Copyright 1999-2018 by Roger Bowler, Jan Jaeger, and others
HHC01415I Build date: Dec  8 2018 at 13:35:57
HHC00403I 0:0000 CKD file C:/Hercules/ZOS23/DASD/a3prd3.cckd: opening as r/o
HHC00414I 0:0000 CKD file C:/Hercules/ZOS23/DASD/a3prd3.cckd: cyls 10017 heads 15 tracks 150255 trklen 56832
HHC00472I 0:0000 CKD64 file C:\Hercules\ZOS23\DASD64\a3prd3.cckd: creating 3390 volume : 10017 cyls, 15 trks/cyl, 56832 bytes/track
HHC00471I 0:0000 CKD64 file C:\Hercules\ZOS23\DASD64\a3prd3.cckd: 10017 cylinders successfully written
HHC00470I 0:0001 CKD64 file C:/Hercules/ZOS23/DASD64/a3prd3.cckd: cyls 10017 heads 15 tracks 150255 trklen 56832
/   0%     185HHC90020W 'hthread_setschedparam()' failed at loc=cckddasd.c:1705: rc=22: Invalid argument
HHC90020W 'hthread_setschedparam()' failed at loc=cckddasd.c:1705: rc=22: Invalid argument
HHC02423I DASD operation completed

```
 However, I think I found a bug.

z/OS 2.3 DASD configuration entries need the additional parameter

`cu=3990-3`

Why this is so I do not really know, all I know is that this is a hard requirement; it won't IPL otherwise. But in the case of CCKD64, I don't even get that far. At Hercules startup, I get HHC00402E error messages, as if the "cu" parameter is not recognized : 

```
16:56:42 HHC01437I Config file[212] hercules.cnf: including file hercules_zos2_dasd.cnf
16:56:42 HHC01603I *0A80 3390 C:\Hercules/ZOS23/DASD64/a3res1.cckd           sf=C:\Hercules/ZOS23/DASD64/SHADOW/HD/a3res1_*.cckd            cu=3990-3
16:56:42 HHC00402E 0:0A80 CKD file: parameter cu in argument 3 is invalid
16:56:42 HHC01463E 0:0A80 device initialization failed
16:56:42 HHC00402E 0:0A81 CKD file: parameter cu in argument 3 is invalid
16:56:42 HHC01463E 0:0A81 device initialization failed
```

I then looked at the source differences between ckddasd.c and ckddasd64.c, but could not discover as to why this "cu=" parameter would no longer be found. I'm hoping you may be able to correct his Fish.

Cheers,

Peter

**Comment by Peter-J-Jansen** [comment]
OK, the above bug was fixed by Fish a day ago, so please ignore my previous entry. Thanks Fish !
As a result, z/OS 2.3 runs just fine using CCKD64 converted DASD.  On my newest Windows 10 1809 system, i9-9900K + Samsung 970 EVO SSD,s with NUMCPU=8, this is the performance : 

```
HHC02272I Highest observed MIPS and IO/s rates:
HHC02272I From Wed Dec 12 00:00:00 2018 to Wed Dec 12 21:09:15 2018
HHC02272I MIPS: 1005.27780
HHC02272I IO/s: 5382          
```

Cheers,

Peter

**Comment by Fish-Git** [comment]
_**FYI:**_ I have created a new utility called `'convto64.exe'` to ease converting your CCKD dasd images to the new CCKD64 format.

It supports converting _base_ dasd images _**or**_ _shadow files_, separately. This means you can use it to convert _only_ a shadow file to the new format _**or**_ _only_ a _base_ dasd image file to the new format. It does _not_ do both at the same time, nor does it merge the shadow file into the base like `dasdcopy64` does. (*)

_It is also **much** faster than dasdcopy64 too!_

 In my timing tests it was able to convert a _**2.38 GB**_ base cckd dasd image file to cckd64 format in _**26 seconds**_, whereas `dasdcopy64` took over _**4 minutes and 19 seconds**_ to accomplish the same thing! &nbsp; _(That's over **10 times faster**, folks!)_

I have used it to convert all of my existing base and shadow files for all of my test Hercules systems (DOS/VS, VM/370, z/OS 1.10, z/OS 2.1, z/OS 2.3 and z/VM 6.3) to the new cckd64 format and everything works beautifully.

Enjoy!&nbsp;`:)`

&nbsp;
_____
(*) `dasdcopy`/`dasdcopy64` requires a _BASE_ cckd dasd image file as input along with an optional shadow file, resulting in a new _BASE_ cckd dasd image file with the optional shadow file automatically merged into it, but it _cannot_ copy _ONLY_ an existing cckd format _shadow file_ to the new cckd64 format.  `convto64` however, is able to do this. It just converts from cckd format to cckd64 format and nothing else. It doesn't do any merging or anything. It's very simple and very fast.

**Comment by Fish-Git** [comment]
_**FIRST POTENTIAL SERIOUS(?) CCKD64 BUG DISCOVERED:**_

Folks, I apologize sincerely for this bug. It's been in there since CCKD64 was first released. While the impact of the bug appears to be serious, the likelihood of it occurring appears to be relatively low. &nbsp;_(I hope!)_

The bug is fixed by commits 1f290ba313a96f6b3bd3a1cf74a9613ffe045469 and 3f35e203ad729a45fb8e7905752ba7f914f97d9c:

**Buggy code:**
```C
DLL_EXPORT int cckd64_chkdsk(DEVBLK *dev, int level)

     [...]

    /*---------------------------------------------------------------
     * Header checks
     *---------------------------------------------------------------*/

     [...]

    /* Device header checks */

     [...]

    if (imgtyp & ANY32_SF_TYP)
        shadow = 0xff;

     [...much later...]

        /* Get storage for the rebuilt l2 tables */
        len = CCKD64_L2TAB_SIZE;
        for (i = 0; i < trks; i++)
        {
            L1idx = (int)(i / 256);
            if (rcvtab[i] != 0 && l2[L1idx] == NULL)
            {
                if ((l2[L1idx] = malloc((size_t)len)) == NULL)
                    goto cdsk_malloc_error;
                l1[L1idx] = shadow ? 0xffffffff : 0;
                memcpy( l2[L1idx], &empty_l2, (size_t)len );
            }
        }
```
&nbsp;
**Corrected code:**
```C
DLL_EXPORT int cckd64_chkdsk(DEVBLK *dev, int level)

     [...]

    /*---------------------------------------------------------------
     * Header checks
     *---------------------------------------------------------------*/

     [...]

    /* Device header checks */

     [...]

    if (imgtyp & ANY64_SF_TYP)
        shadow = 0xff;

     [...much later...]

        /* Get storage for the rebuilt l2 tables */
        len = CCKD64_L2TAB_SIZE;
        for (i = 0; i < trks; i++)
        {
            L1idx = (int)(i / 256);
            if (rcvtab[i] != 0 && l2[L1idx] == NULL)
            {
                if ((l2[L1idx] = malloc((size_t)len)) == NULL)
                    goto cdsk_malloc_error;
                l1[L1idx] = shadow ? ULLONG_MAX : 0;
                memcpy( l2[L1idx], &empty_l2, (size_t)len );
            }
        }
```

&nbsp;
**As best as I can tell** _(the cckd checkdisk logic is rather complicated and hard to follow!)_**,** the bug would only occur when a track was being recovered (`rcvtab[i] != 0`) on a cckd64 image.

So as long as there was no existing damage to any of your existing cckd64 files (i.e. as long as you have never had the need to run `cckdcdsk64` or the `sfk` command on any of your cckd64 dasd), then the bug would not occur. It appears to only be triggered by running `cckdcdsk64` (or the `sfk` command) for a cckd64 dasd image. (And then only under certain difficult-to-determine conditions.)

I sincerely hope this bug hasn't impacted any of you and sincerely apologize if it has. I don't know how I missed it. I've probably desk checked my code at least a dozen times by now, and yet I  somehow completely missed it each time.

I am currently in the process of desk checking each of the `....64.c` source files _yet again_ (for the 47 millionth time) just in case anything else jumps out at me.

Thanks.

_(and sorry about this!)_

p.s. How's everything going otherwise?&nbsp;`:)`

**Comment by Fish-Git** [comment]
Hmmmm...

It looks like the impact of my bug could be a bit larger than originally thought.

During my desk checking (still in progress) I discovered that the dasd device handler initialization function (`cckd64_dasd_init_handler`) _always calls the check-disk function_ `cckd64_chkdsk` during initialization, just as a matter of routine (line 71 of source file `cckddasd64.c`).

_**However**,_ the good(?) news is, since it always calls it requesting only a _level-0_ check (devhdr, cdevhdr, l1 table, l2 tables only), it looks like the previously mentioned bug will probably _not_ be triggered. The bug only appears to _only_ (_maybe_) be triggered if a track is being recovered (previously mentioned `rcvtab[i] != 0`), and a level-0 check-disk does not recover track images. Only a level-3 or level-4 check-disk will attempt to recover track images.

So I think we're safe. &nbsp; _(Whew!)_

**Comment by Fish-Git** [comment]
Reminder for self _(or someone else willing to take up the challenge!)_:

- [ ] Update/fix `cckddump.hla` to support new CCKD64 Hercules dasd image file format!
- [ ] Update/fix `cckdload.hla` to support new CCKD64 Hercules dasd image file format!

Both are HLASM assembler source files living in the repository's "util" subdirectory. They were written by Greg Smith (the original author/designer of our CKD/CCKD emulated dasd support) and later updated by Chistopher Varlet in 2014 to support "very large dasds" (i.e. those with more than 64K tracks) and are designed to allow you to transfer a mainframe dasd volume from your mainframe to your PC for use with Hercules.&nbsp;&nbsp;(and vice versa?)

_**HOWEVER...**_ Even though the cckddump/cckdload programs both now support dasd device types with more than 64K tracks, they both only support the original 32-bit CCKD (Compressed CKD) file format (presumably(*)).

If a large dasd volume cannot be _compressed_ into a Hercules CCKD dasd image file of _less_ than 4GB in size, then you're basically screwed. You can't transfer that volume from your mainframe over to Hercules on your PC (because the size of the resulting CCKD file would be larger than the 4GB limit).

If the very large mainframe dasd volume had very little data on it however (so that it could be compressed to _less_ than 4GB), then `cckddump` and `cckdload` can be used just fine and Hercules can use the resulting CCKD dasd image files just fine too.

But usually very large dasds (with lots of tracks) are usually very large for a good reason! Such dasd volumes usually contain _lots_ of data! _So much_ data that all of it cannot be compressed down to less than 4GB! And if the _compressed_ size of the Hercules dasd image file is greater than 4GB, you're basically out of luck.

The new CCKD64 dasd image file format fixes this shortcoming of course, so that now your CCKD64 dasd image files can be as large as you need them to be and they'll work just fine with SDL Hyperion.

_**BUT..**_ What about `cckddump` and `cckdload`?

They are both probably(**) still creating their files in the old 32-bit CCKD format and not the new CCKD64 file format.

Thus the need for this reminder for myself (or someone else!) to review both of these programs and make whatever changes might be needed to have them create their files in the new CCKD64 file format instead.
___
(*) I have not yet had the chance to closely review either one to determine whether the problem actually exists or not, but I am presuming it does.
(**) Ibid.

**Comment by Fish-Git** [comment]
**FYI:** a new GitHub Issue (#167 "[CCKD64 support needed for cckddump/cckdload utils](#167)") has been created as a sub-issue (sub-project) of this one to address the cckddump/cckdload mainframe assembler utilities that were missed in the original CCKD64 effort.

**Comment by Fish-Git** [comment]
I am going to go ahead and close this issue at this time just to get it off my list. We can deal with whatever individual CCKD64 issues that may come up separately, but the basic issue of CCKD64 not existing has been completed. CCKD64 support now exists.
